lib/hostip.c

```diff
@@ -819,48 +819,95 @@ CURLcode Curl_loadhostpairs(struct Curl_easy *data)
     }
     else {
       struct Curl_dns_entry *dns;
-      Curl_addrinfo *addr;
+      Curl_addrinfo *head = 0, *tail = 0;
       char *entry_id;
       size_t entry_len;
-      char buffer[256];
-      char *address = &buffer[0];
-
-      if(3 != sscanf(hostp->data, "%255[^:]:%d:%255s", hostname, &port,
-                     address)) {
-        infof(data, "Couldn't parse CURLOPT_RESOLVE entry '%s'!\n",
-              hostp->data);
-        continue;
+      char address[64];
+      char *addrPointer;
+      char *last;
+      char *tmp;
+      unsigned long tmpPort;
+      int error = true;
+
+      tmp = strchr(hostp->data, ':');
+      if(!tmp || (tmp - hostp->data) >= (ptrdiff_t)sizeof(hostname)) {
+        goto err;
       }
+      memcpy(hostname, hostp->data, tmp - hostp->data);
+      hostname[tmp - hostp->data] = '\0';
 
-      /* allow IP(v6) address within [brackets] */
-      if(address[0] == '[') {
-        size_t alen = strlen(address);
-        if(address[alen-1] != ']')
-          /* it needs to also end with ] to be valid */
-          continue;
-        address[alen-1] = 0; /* zero terminate there */
-        address++; /* pass the open bracket */
+      tmpPort = strtoul(tmp + 1, &tmp, 10);
+      if(!tmpPort || tmpPort > USHRT_MAX || *tmp != ':' || !(tmp + 1)) {
```

Comment:
I think you meant !*(tmp+1)

Reply:
also curl supports port 0 [or so I've read](https://daniel.haxx.se/blog/2014/10/25/pretending-port-zero-is-a-normal-one/)

Reply:
I did indeed :-) I also removed the !tmpPort check and rather added a check for the no valid characters case.